### ISSUE TITLE
get

### ISSUE THREAD

**Issue by macartan** [integrated-inferences/CausalQueries#319]
1. parameter distribution: we have a redundnancy between:
get_param_dist
get_prior_distribution
get_posterior_distribution
get_param_dist is used internally and probably should remain but all of these might be replaced by:
get(object = "parameter_priors")
get(object = "parameter_posteriors")

2. causal type distribution
prior and posteriors on the type distribution  can be accessed using
get_type_prob_multiple
This should be internal (probably get_type_prob also)
get(object = "types_priors")
get(object = "types_posteriors")

**Comment by macartan** [comment]
NOTE ALSO:
get_posterior() return a dataframe with a column for each type
get_type_prob_multiple(base_0_1, using = "posteriors") returns a matrix with a row for each type!

I think only the matrices are used internally whereas the datasets are the natural user facing object

**Comment by till-tietz** [comment]
Copying email thread discussion on generic `get` method here for visibility: 

generic `get` could look something like this:

we already have get_* functions for a lot of model objects (these are used quite heavily internally by other functions): we make them internal and assign their output a class e.g. for nodal_types output of get_nodal_types() will be of class "nodal_types" with an accompanying print method print.nodal_types(). We then call internal get_* methods in the generic get method get.causal_model() attached to the "causal_model" class.

get.causal_model <- function(x, object) {
     if (object == "nodal_types") {
         return(get_nodal_types(x))
     }

     ....
}

The workflow for inspecting model objects could then look like this:

model |> get("nodal_types") |> print()

**Comment by macartan** [comment]
Looks good to me;
here:
model |> get("nodal_types") |> print()

one wouldn't need the `print()` at the end, right?

we should have tests that cover cases where the object does not exist; e.g.
get_posterior_distribution() will handle cases where the posterior does or
does not exist

**Comment by till-tietz** [comment]
@macartan @lilymedina @gerasy1987 

I'm just writing the print method for all the causal_model objects. Wanted to get your feedback on how we want the output of print for type / parameter distributions to look: 

I was thinking something like this: 

```
Distribution Name:

                     Min.     1st Qu.     Median       Mean    3rd Qu.      Max.
  X0.M00.Y00 2.579687e-06 0.005937542 0.01582735 0.02237553 0.03313530 0.1384651
  X1.M00.Y00 1.417390e-05 0.006356912 0.01654490 0.02149115 0.03180133 0.1196712
  X0.M10.Y00 1.766622e-05 0.006295554 0.01590068 0.02208618 0.03238188 0.1481497
  X1.M10.Y00 1.061271e-05 0.006482361 0.01558554 0.02129570 0.03182914 0.1146359
  X0.M01.Y00 1.234858e-06 0.009249786 0.02166560 0.02806984 0.04144582 0.1575391
  X1.M01.Y00 2.386358e-06 0.009571449 0.02194888 0.02714218 0.03920952 0.1494898
  X0.M11.Y00 3.878263e-05 0.009446221 0.02189427 0.02818947 0.04127971 0.1357763
  X1.M11.Y00 4.650882e-05 0.009412006 0.02168522 0.02716645 0.03910148 0.1434938
  X0.M00.Y10 1.110632e-05 0.009269398 0.02240432 0.02846005 0.04169611 0.1664499
  X1.M00.Y10 3.743552e-05 0.010376407 0.02326374 0.02888756 0.04250640 0.1318902
  X0.M10.Y10 3.654964e-06 0.009464590 0.02194131 0.02833778 0.04160753 0.1672520
  X1.M10.Y10 7.282908e-06 0.010488442 0.02426038 0.02928555 0.04246260 0.1458881
  X0.M01.Y10 2.409971e-06 0.013993510 0.02966312 0.03597952 0.05141401 0.1938576
  X1.M01.Y10 9.329095e-07 0.016063540 0.03155166 0.03692801 0.05300211 0.1408780
  X0.M11.Y10 1.202570e-05 0.013560196 0.02984301 0.03554062 0.05135620 0.1813975
  X1.M11.Y10 8.658909e-05 0.015462251 0.03141939 0.03641696 0.05215857 0.1489689
  X0.M00.Y01 5.525468e-06 0.009350166 0.02199537 0.02806370 0.04059289 0.1619948
  X1.M00.Y01 6.595926e-07 0.006389762 0.01600728 0.02130809 0.03135581 0.1289643
  X0.M10.Y01 3.455777e-06 0.009199942 0.02176246 0.02794458 0.04104255 0.1763842
  X1.M10.Y01 2.292192e-06 0.006280676 0.01610603 0.02135735 0.03139991 0.1193510
  X0.M01.Y01 1.780783e-06 0.013633880 0.02940115 0.03545736 0.05104083 0.1890317
  X1.M01.Y01 2.140939e-06 0.009543773 0.02169027 0.02714591 0.03985708 0.1587937
  X0.M11.Y01 1.017211e-05 0.013646656 0.02940009 0.03525636 0.05099165 0.1708749
  X1.M11.Y01 3.783790e-06 0.009368671 0.02195073 0.02688429 0.03924610 0.1328016
  X0.M00.Y11 7.265099e-05 0.018246973 0.03862171 0.04464463 0.06538300 0.2010720
  X1.M00.Y11 1.459762e-05 0.008595768 0.02040040 0.02580964 0.03820757 0.1346555
  X0.M10.Y11 1.182361e-04 0.019119309 0.03752205 0.04445248 0.06376732 0.1959237
  X1.M10.Y11 4.980382e-06 0.008967908 0.01989906 0.02573128 0.03724185 0.1278635
  X0.M01.Y11 5.361784e-06 0.027434828 0.04985150 0.05649142 0.07916275 0.2026794
  X1.M01.Y11 4.033814e-06 0.013435925 0.02698320 0.03271963 0.04710385 0.1526324
  X0.M11.Y11 1.627131e-04 0.026807208 0.05086016 0.05644727 0.08030783 0.2214116
  X1.M11.Y11 5.708975e-05 0.012795920 0.02701877 0.03263346 0.04733460 0.1589160
```
maybe an extra line of description on what's going on would also be nice. Let me know what you think

**Comment by macartan** [comment]
I saw in lilys branch I think that get used switch and employed existing
get_ functions when they exist

That looked right to me and in that case I’d suggest making all the other
get_ functions internal

**Comment by macartan** [comment]
For compactness I’d suggest mean and ci, would be enough and the round to a
small number of decimal places (though if many don’t want all 0s so
benchmark off the largest values perhaps ?)

Maybe also just first 10 rows if large?

Plus a description giving dimension ? And meanings of rows and columns ?

**Comment by lilymedina** [comment]
I made changes in the new "grab" branch, not in "get". There were conflicts because of recent edits in the master branch that I couldn't resolve so I created a new branch :/. Sorry for the mess. I kept the "get" branch just in case someone else was using it.

Also, I renamed the base "get" function to "grab" because it was causing trouble when updating the model. The base "get" function is used in "stanmodels.R".

I made internal all the "get" functions that get used to get any of the core elements of a causal_model; there are still other get functions that gets exported e.g get_ambiguities_matrix.

Also I deprecated  the "get_prior_distribution" and "get_posterior_distribution" functions.

**Comment by till-tietz** [comment]
Fantastic!!! Thank you @lilymedina

**Comment by macartan** [comment]
Why not include the get ambiguities matrix in grab and stop exporting it?

**Comment by till-tietz** [comment]
@macartan @lilymedina @gerasy1987 

classes and S3 method are implemented for: 

- dag
- statement 
- nodes
- parents_df
- parameters_df
- causal_types
- nodal_types 
- parameter_matrix
- parameter_prior
- parameter_posterior 
- parameters

will implement 
- type_posterior 
- type_prior
- even_probabilites
- stan_fit 

shortly 

One question is whether we want/need a print method for stan_object

**Comment by macartan** [comment]
If people save the original stan object it comes with its own print method
from stan?
t

**Comment by gerasy1987** [comment]
@macartan yes, this is how the stan object saved in causal_model (if `keep_stanfit = TRUE`) is printed at the moment

**Comment by lilymedina** [comment]
@macartan the only reason I didn't do `get_ambiguity_matrix` and some other functions is because we had discussed using the `grab` function for the objects that were appended to a `causal_model`. I can easily include the `get_ambiguity_matrix` and other functions into the `grab` function as well.

**Comment by macartan** [comment]
My sense is we should use it for anything that can be inferred directly
from the model

get parmap also, get event probabilities

I think this will make it easier for users to figure out everything they
can do?


But not things that vary like make_data

Upside is that one help file lets people
See all the options together. But downside maybe is that help file is long
if we want to give many examples for some objects
